File: src/ast/FoldExpression.java
package ast;

import java.io.StringWriter;

public class FoldExpression extends BaseNode {
    private boolean _isLeftFold;
    private String _operatorName;
    private BaseNode _expression;
    private BaseNode _initializer;

    public FoldExpression(boolean isLeftFold, String operatorName, BaseNode expression, BaseNode initializer) {
        super(NodeType.FunctionParameter);
        _isLeftFold = isLeftFold;
        _operatorName = operatorName;
        _expression = expression;
        _initializer = initializer;
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        writer.write("(");

        if (_isLeftFold && _initializer != null) {
            _initializer.Print(writer);
            writer.write(" ");
            writer.write(_operatorName);
            writer.write(" ");
        }

        writer.write(_isLeftFold ? "... " : " ");
        writer.write(_operatorName);
        writer.write(!_isLeftFold ? " ..." : " ");
        _expression.Print(writer);

        if (!_isLeftFold && _initializer != null) {
            _initializer.Print(writer);
            writer.write(" ");
            writer.write(_operatorName);
            writer.write(" ");
        }

        writer.write(")");
    }
}

File: src/de/fearlesstobi/demangler/ast/FoldExpression.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class FoldExpression(private val isLeftFold: Boolean, private val operatorName: String, private val expression: BaseNode, private val initializer: BaseNode?) : BaseNode(NodeType.FunctionParameter) {
    public override fun printLeft(writer: StringWriter) {
        writer.write("(")
        if (isLeftFold && initializer != null) {
            initializer.print(writer)
            writer.write(" ")
            writer.write(operatorName)
            writer.write(" ")
        }
        writer.write(if (isLeftFold) "... " else " ")
        writer.write(operatorName)
        writer.write(if (!isLeftFold) " ..." else " ")
        expression.print(writer)
        if (!isLeftFold && initializer != null) {
            initializer.print(writer)
            writer.write(" ")
            writer.write(operatorName)
            writer.write(" ")
        }
        writer.write(")")
    }

}